### ISSUE TITLE
feature request: manage topics via admin API

### ISSUE THREAD

**Issue by bufferoverflow** [gitlabhq/terraform-provider-gitlab#748]
With the upcoming GitLab release there is an [admin API to manage topics](https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/api/topics.md), see also https://gitlab.com/gitlab-org/gitlab/-/issues/340920

Would be awesome to become able to manage the topics via the terraform provider.

**Comment by timofurrer** [comment]
I've created another issue in the go-gitlab library to support the API so that it can easily implemented here. see https://github.com/xanzy/go-gitlab/issues/1288

**Comment by eliobischof** [comment]
As the GitLab API doesn't support deleting a topic, implementing a fully functional topic resource isn't possible ATM.

That's what the docs say about resource implementation...

```golang
	// The functions below are the CRUD operations for this resource.
	//
	// The only optional operation is Update. If Update is not implemented,
	// then updates will not be supported for this resource.
	//
	// The ResourceData parameter in the functions below are used to
	// query configuration and changes for the resource as well as to set
	// the ID, computed data, etc.
	//
	// The interface{} parameter is the result of the ConfigureFunc in
	// the provider for this resource. If the provider does not define
	// a ConfigureFunc, this will be nil. This parameter should be used
	// to store API clients, configuration structures, etc.
	//
	// If any errors occur during each of the operation, an error should be
	// returned. If a resource was partially updated, be careful to enable
	// partial state mode for ResourceData and use it accordingly.
	//
	// Exists is a function that is called to check if a resource still
	// exists. If this returns false, then this will affect the diff
	// accordingly. If this function isn't set, it will not be called. You
	// can also signal existence in the Read method by calling d.SetId("")
	// if the Resource is no longer present and should be removed from state.
	// The *ResourceData passed to Exists should _not_ be modified.
	Create CreateFunc
	Read   ReadFunc
	Update UpdateFunc
	Delete DeleteFunc
	Exists ExistsFunc
```

_The only optional operation is Update._

Of course I can pass a delete function that always returns an error saying that deleting a topic is not supported...

**Comment by bufferoverflow** [comment]
@eliobischof for the delete case the option I see at the moment is mapping it to update endpoint and just remove the image and empty the description of the topic. Anyway, first we need https://github.com/xanzy/go-gitlab/issues/1288

**Comment by eliobischof** [comment]
@bufferoverflow I can empty the description but not the avatar, as the update API requires a valid avatar URL

**Comment by bufferoverflow** [comment]
@wwwjon Could you please check and maybe extend the topics API to allow deletion of an avatar so we get this done here for the self-service?

@eliobischof just proceed with a feasible workaround until GitLab can handle this

**Comment by wwwjon** [comment]
**Delete an avatar:**
Indeed, it seems that you can only delete an avatar via the UI and not via API (the same applies to avatars of projects, groups and users). I will tackle this so that it will be possible via API as well.

**Delete a topic:**
It was decided (in a first version) that topics cannot be deleted, because users can create any topic again and again.

My personal opinion is that it's confusing when you want to delete a topic and instead only its avatar and description are cleared.  That's why I think the solution _"return an error saying that deleting a topic is not supported"_ is better for now, until the deletion of a topic via API might be supported later. The change from "clear properties only" to "delete completely" is a real API change, while the change from "not supported" to "correctly supported" is simply extending the functionality.

**Comment by wwwjon** [comment]
> I have just submitted a [Merge Request on GitLab](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/76860) to allow to remove the avatar by update via Topics API using a blank parameter value:

The merge request is merged on GitLab and the new feature should now be available on gitlab.com (see [documentation](https://docs.gitlab.com/ee/api/topics.html#remove-a-topic-avatar))

/cc @eliobischof @bufferoverflow

**Comment by armsnyder** [comment]
closed via #871 , will be in the upcoming 3.10.0 release.
